src/main/java/com/epam/aidial/core/service/ResourceService.java

```diff
@@ -0,0 +1,401 @@
+            }
+        }
+
+        return new Result(body, createdAt, updatedAt, true, true);
+    }
+
+    private void blobPut(String key, String body, long createdAt, long updatedAt) {
+        byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
+        String encoding = null;
+
+        if (bytes.length >= compressionMinSize) {
+            encoding = "gzip";
+            bytes = Compression.compress(encoding, bytes);
+        }
+
+        Map<String, String> metadata = Map.of(
+                "created_at", Long.toString(createdAt),
+                "updated_at", Long.toString(updatedAt)
+        );
+        blobStore.store(key, "application/json", encoding, metadata, bytes);
+    }
+
+    private void blobDelete(String key) {
+        blobStore.delete(key);
+    }
+
+    private static String blobKey(ResourceDescription descriptor) {
+        String path = descriptor.getAbsoluteFilePath();
+        return descriptor.isFolder() ? path : (path + BLOB_EXTENSION);
+    }
+
+    private static String blobKeyFromRedisKey(String redisKey) {
+        int i = redisKey.indexOf(":");
+        return redisKey.substring(i + 1) + BLOB_EXTENSION;
+    }
+
+    private static String fromBlobKey(String blobKey) {
+        return blobKey.endsWith(BLOB_EXTENSION) ? blobKey.substring(0, blobKey.length() - BLOB_EXTENSION.length()) : blobKey;
+    }
+
+    @Nullable
+    private Result redisGet(String key, boolean withBody) {
+        RMap<String, String> map = redis.getMap(key, StringCodec.INSTANCE);
+        Map<String, String> fields = map.getAll(withBody ? REDIS_FIELDS : REDIS_FIELDS_NO_BODY);
+
+        if (fields.isEmpty()) {
+            return null;
+        }
+
+        String body = fields.get("body");
+        long createdAt = Long.parseLong(fields.get("created_at"));
+        long updatedAt = Long.parseLong(fields.get("updated_at"));
+        boolean synced = Boolean.parseBoolean(Objects.requireNonNull(fields.get("synced")));
+        boolean exists = Boolean.parseBoolean(Objects.requireNonNull(fields.get("exists")));
+
+        return new Result(body, createdAt, updatedAt, synced, exists);
+    }
+
+    private void redisPut(String key, Result result) {
+        RScoredSortedSet<String> set = redis.getScoredSortedSet(REDIS_QUEUE, StringCodec.INSTANCE);
+        set.add(time() + syncDelay, key);
+
+        RMap<String, String> map = redis.getMap(key, StringCodec.INSTANCE);
+
+        if (!result.synced) {
+            map.clearExpire();
+        }
+
+        Map<String, String> fields = Map.of(
+                "body", result.body,
+                "created_at", Long.toString(result.createdAt),
+                "updated_at", Long.toString(result.updatedAt),
+                "synced", Boolean.toString(result.synced),
+                "exists", Boolean.toString(result.exists)
+        );
+        map.putAll(fields);
+
+        if (result.synced) {
+            map.expire(cacheExpiration);
+            set.remove(key);
```

Comment:
We add and the remove the same key in the same method. Can you please add one line comment why is so?